### PR TITLE
feat(1710): support external join, rewrite trigger logic [5]

### PR DESCRIPTION
**Pull request by d2lam**
To support external join, we need to rewrite the whole trigger logic. 

1. workflow graph now includes both internal & external relationship (logic in https://github.com/screwdriver-cd/workflow-parser/pull/27/files)
2. we can get the `joinList` from workflow graph (logic in https://github.com/screwdriver-cd/workflow-parser/pull/27/files)
3. after a job is done, find out all the next jobs it triggers. For each job:
   * find its joined builds (for example next job is D, and it requires `[A,sd@111:B, sd@222:C]`
   * construct a `parentBuilds` field = current build's parentBuilds + all jobs in `joinList` + current build info. Looks like this: `{ 999: {eventId: 1, A: 1}, 111: {eventId: 2, B: 2}, 222: {eventId: null, C:null } }`
   * if this list is empty (this job only requires `or`), create/start build D (with `parentBuilds` info)
   * if this list is not empty (this job requires `and`)
       * check if build D already exists
           * if internal, check by looking at builds inside current event/parent event 
           * if external, check by looking at the latest build of D and see if status is `CREATED`
       * if build D doesn't exist, create build D (with `parentBuilds` info), don't start D yet. 
       * once build D exists, check if `[A,sd@111:B, sd@222:C]` are done by looking at its `parentBuilds` field
       * if has one failure, delete build D
       * if some build in `joinList` is not done, do nothing
       * if all finished successfully, start build D

Related PRs:
- https://github.com/screwdriver-cd/data-schema/pull/361
- https://github.com/screwdriver-cd/models/pull/400
- https://github.com/screwdriver-cd/config-parser/pull/94
- https://github.com/screwdriver-cd/workflow-parser/pull/27

Related Issue: https://github.com/screwdriver-cd/screwdriver/issues/1710

### CODE REVIEW

File: plugins/builds/index.js [hunk 1]
```diff
@@ -11,10 +11,98 @@ const listSecretsRoute = require('./listSecrets');
...
 const workflowParser = require('screwdriver-workflow-parser');
+const deepmerge = require('deepmerge');
+const schema = require('screwdriver-data-schema');
+const { EXTERNAL_TRIGGER_AND } = schema.config.regex;
 
 /**
- * Create the build. If config.start is false or not passed in then do not start the job
- * @method createBuild
+ * Create event for downstream pipeline that need to be rebuilt
+ * @method createEvent
```
Comment (d2lam) [inline plugins/builds/index.js]: this function is the same as previous `triggerEvent`
factored it out since we need to call it somewhere else.

File: plugins/builds/index.js [hunk 2]
```diff
@@ -293,83 +410,329 @@ exports.register = (server, options, next) => {
...
+            let nextBuild;
+
+            /* CHECK WHETHER NEXT BUILD EXISTS */
+            // if next build is external, check the latest build
+            if (isExternal) {
+                const p = await pipelineFactory.get(externalPipelineId);
+                const jobArray = await p.getJobs({ params: { name: externalJobName } });
+                const j = await jobFactory.get(jobArray[0].id);
+
+                nextBuild = await j.getLatestBuild({ status: 'CREATED' });
```
Comment (d2lam) [inline plugins/builds/index.js]: per @jithin1987  suggestion, we should store a new field maybe `upstreamEvent`, to trace back to previous event. So that we can get the actual build instead of just fetching the latest build here

File: plugins/builds/index.js [hunk 3]
```diff
@@ -293,83 +410,329 @@ exports.register = (server, options, next) => {
...
+                jName = name;
+            }
+
+            return { pId, jName };
+        };
+
+        console.log(joinObj);
+
+        // Go through each next job
+        return Promise.all(Object.keys(joinObj).map(async (nextJobName) => {
```
Comment (d2lam) [inline plugins/builds/index.js]: Still need to handle restart case. For example
join list of `D` is `[A, sd@111:B, sd@222:C]`
`A and sd@222:C` passed but `sd@111:B` failed
if we restart `sd@111:B`, then `D` should know to fetch previous successful info from `A and sd@222:C` build. This will require additional logic

File: test/plugins/builds.test.js [hunk 4]
```diff
@@ -1754,6 +1838,471 @@ describe('build plugin test', () => {
...
+
+                    buildFactoryMock.get.withArgs(5555).resolves({ status: 'FAILURE' });
+
+                    return newServer.inject(options).then(() => {
+                        assert.notCalled(buildFactoryMock.create);
+                        assert.calledOnce(buildC.remove);
+                    });
+                });
+
+                it.skip('triggers if all jobs in join are done with parent event', () => {
```
Comment (d2lam) [inline test/plugins/builds.test.js]: This is the test for restart

File: test/plugins/builds.test.js [hunk 5]
```diff
@@ -1754,6 +1838,471 @@ describe('build plugin test', () => {
                     });
                 });
             });
+
+            describe('join new flow', () => {
```
Comment (d2lam) [inline test/plugins/builds.test.js]: this is the tests for the new flow
current added tests are for internal join only
need to add more tests for external join workflow

File: plugins/builds/index.js [hunk 6]
```diff
@@ -158,6 +180,137 @@ function handleNextBuild({ buildConfig, joinList, finishedBuilds, jobId }) {
...
+ * @param {String}  config.startFrom        Job to be rebuilt
+ * @param {String}  config.causeMessage     Caused message, e.g. triggered by 1234(buildId)
+ * @param {String}  config.parentBuildId    ID of the build that triggers this event
+ * @param {Object} [config.parentBuilds]    Builds that triggered this build
+ * @return {Promise}                        Resolves to the newly created event
+ */
+async function createEvent(config) {
+    const { pipelineFactory, eventFactory, pipelineId, startFrom,
+        causeMessage, parentBuildId, parentBuilds } = config;
+    const scm = eventFactory.scm;
```
Comment (tkyi) [inline plugins/builds/index.js]:
```suggestion
    const { scm } = eventFactory;
```
🤷‍♀

Comment (d2lam) [reply plugins/builds/index.js]: 👍  feel free to amend

File: plugins/builds/index.js [hunk 7]
```diff
@@ -158,6 +161,137 @@ function handleNextBuild({ buildConfig, joinList, finishedBuilds, jobId }) {
...
+ */
+async function createInternalBuild(config) {
+    const { jobFactory, buildFactory, eventFactory, pipelineId, jobName,
+        username, scmContext, build, parentBuilds, start, baseBranch } = config;
+    const event = await eventFactory.get(build.eventId);
+    const job = await jobFactory.get({
+        name: jobName,
+        pipelineId
+    });
+    const prRef = event.pr.ref ? event.pr.ref : '';
```
Comment (tkyi) [inline plugins/builds/index.js]:
```suggestion
    const prRef = event.pr.ref || '';
```
same thing

Comment (tkyi) [reply plugins/builds/index.js]: might want to do `hoek.reach` though

Comment (d2lam) [reply plugins/builds/index.js]: yeah, can change it but this is the old code from here: https://github.com/screwdriver-cd/screwdriver/blob/master/plugins/builds/index.js#L31-L56

File: plugins/builds/index.js [hunk 8]
```diff
@@ -293,83 +391,329 @@ exports.register = (server, options, next) => {
...
+
+            // join jobs, with eventId and buildId empty
+            const joinParentBuilds = {};
+
+            joinListNames.forEach((name) => {
+                const { pId, jName } = getPipelineAndJob(name);
+
+                joinParentBuilds[pId] = {
+                    eventId: null,
+                    [jName]: null
```
Comment (tkyi) [inline plugins/builds/index.js]:
```suggestion
                   jobs: [{ jName: null }]
```

File: plugins/builds/index.js [hunk 9]
```diff
@@ -293,83 +391,329 @@ exports.register = (server, options, next) => {
...
                 };
+            });
+            // need to deepmerge because it's possible same event has multiple builds
+            // const combined = deepmerge(currentJobParentBuilds, joinParentBuilds);
+
+            // override currentBuild in the joinParentBuilds
+            const currentBuildInfo = {
+                [pipelineId]: {
+                    eventId: build.eventId,
+                    [currentJobName]: build.id
```
Comment (tkyi) [inline plugins/builds/index.js]:
```suggestion
                    jobs: { currentJobName: build.id }
```

File: plugins/builds/index.js [hunk 10]
```diff
@@ -293,83 +391,329 @@ exports.register = (server, options, next) => {
...
+        // New implementation that allows external join
+        const pipelineFactory = server.root.app.pipelineFactory;
+        const event = await eventFactory.get({ id: build.eventId });
+        const workflowGraph = event.workflowGraph;
+        const nextJobs = workflowParser.getNextJobs(workflowGraph,
+            { trigger: currentJobName, chainPR: pipeline.chainPR });
+        // Create a join object like: {A:[B,C], D:[B,F]} where [B,C] join on A, [B,F] join on D, etc.
+        // This can include external jobs
+        const joinObj = nextJobs.reduce((obj, jobName) => {
+            obj[jobName] = workflowParser.getSrcForJoin(workflowGraph, { jobName });
```
Comment (tkyi) [inline plugins/builds/index.js]: double check here for external build
`[~D, B,C] --> sd@123:A`
Then joinList of `sd@123:A` is `[sd@111:B, sd@111:C]`

File: plugins/builds/index.js [hunk 11]
```diff
@@ -293,83 +391,329 @@ exports.register = (server, options, next) => {
...
+                if (!isExternal) {
+                    return createInternalBuild(internalBuildConfig);
+                }
+
+                return createExternalBuild(externalBuildConfig);
+            }
+
+            let nextBuild;
+
+            /* CHECK WHETHER NEXT BUILD EXISTS */
```
Comment (tkyi) [inline plugins/builds/index.js]: know forsure the current job is in a join list

File: plugins/builds/index.js [hunk 12]
```diff
@@ -293,83 +391,329 @@ exports.register = (server, options, next) => {
...
-                // 2. ([~D,B,C]->A) currentJob=D, nextJob=A, joinList(A)=[B,C]
-                //    joinList doesn't include C, so start A
-                if (joinList.length === 0 || !joinListNames.includes(currentJobName)) {
-                    return createBuild(buildConfig);
+                if (!event.parentEventId) {
+                    finishedInternalBuilds = await event.getBuilds();
+
+                // If parent event id, merge parent build status data and
+                // rerun all builds in the path of the startFrom
+                } else {
```
Comment (d2lam) [inline plugins/builds/index.js]: Similar to old logic here: https://github.com/screwdriver-cd/screwdriver/blob/master/plugins/builds/index.js#L343-L365

File: plugins/builds/index.js [hunk 13]
```diff
@@ -293,83 +391,329 @@ exports.register = (server, options, next) => {
...
+            let hasFailure = false;
+            const promisesToAwait = [];
+
+            for (let i = 0; i < joinListNames.length; i += 1) {
+                const name = joinListNames[i];
+                const { pId, jName } = getPipelineAndJob(name);
+                let bId;
+
+                if (upstream[pId] && upstream[pId][jName]) {
+                    bId = upstream[pId][jName];
```
Comment (tkyi) [inline plugins/builds/index.js]:
```suggestion
                    bId = upstream[pId].jobs[jName];
```

File: plugins/builds/index.js [hunk 14]
```diff
@@ -293,83 +391,329 @@ exports.register = (server, options, next) => {
...
+            let done = true;
+            let hasFailure = false;
+            const promisesToAwait = [];
+
+            for (let i = 0; i < joinListNames.length; i += 1) {
+                const name = joinListNames[i];
+                const { pId, jName } = getPipelineAndJob(name);
+                let bId;
+
+                if (upstream[pId] && upstream[pId][jName]) {
```
Comment (tkyi) [inline plugins/builds/index.js]:
```suggestion
                if (upstream[pId] && upstream[pId].jobs[jName]) {
```

File: plugins/builds/index.js [hunk 15]
```diff
@@ -293,83 +391,329 @@ exports.register = (server, options, next) => {
...
+                else promisesToAwait.push(buildFactory.get(bId));
+            }
+
+            // Check if some joined build failed, so that we can delete next build
+            const joinedBuilds = await Promise.all(promisesToAwait);
+
+            joinedBuilds.forEach((b) => {
+                console.log(b.status);
+                if (b.status === 'FAILURE') hasFailure = true;
+                if (b.status !== 'SUCCESS' && b.status !== 'FAILURE') done = false; // some builds are still going on
```
Comment (tkyi) [inline plugins/builds/index.js]: also need to check UNSTABLE and ABORTED status

File: plugins/builds/update.js [hunk 16]
```diff
@@ -179,37 +181,45 @@ module.exports = () => ({
...
-                                    causeMessage: `Triggered by build ${username}`,
-                                    parentBuildId: newBuild.id
+                            pipeline, job, build: newBuild, username, scmContext, externalJoin
+                        }).then(() => {
+                            // if external join is allow, then triggerNextJobs will take care of external OR already
+                            if (externalJoin) {
+                                return reply(newBuild.toJson()).code(200);
+                            }
+
+                            return triggerFactory.list({ params: { src } })
```
Comment (tkyi) [inline plugins/builds/update.js]: // old flow

File: plugins/builds/index.js [hunk 17]
```diff
@@ -158,6 +180,136 @@ function handleNextBuild({ buildConfig, joinList, finishedBuilds, jobId }) {
...
+
+    return createEvent({
+        pipelineFactory,
+        eventFactory,
+        pipelineId: externalPipelineId,
+        startFrom: externalJobName,
+        parentBuildId, // current build
+        causeMessage,
+        parentBuilds,
+        start: start !== false
```
Comment (MysticDoll) [inline plugins/builds/index.js]: This `start` argument is not needed in `createEvent`.

File: plugins/builds/index.js [hunk 18]
```diff
@@ -158,6 +180,136 @@ function handleNextBuild({ buildConfig, joinList, finishedBuilds, jobId }) {
...
+            jobId: job.id,
+            sha: build.sha,
+            parentBuildId: build.id,
+            parentBuilds,
+            eventId: build.eventId,
+            username,
+            configPipelineSha: event.configPipelineSha,
+            scmContext,
+            prRef,
+            start: start !== false,
```
Comment (MysticDoll) [inline plugins/builds/index.js]:
```suggestion
            start
```

`start !== false` is equivalent to  `start`. How about to use above style?

File: plugins/builds/index.js [hunk 19]
```diff
@@ -293,83 +409,319 @@ exports.register = (server, options, next) => {
...
-                }).then(finishedBuilds => handleNextBuild({
-                    buildConfig,
-                    joinList,
-                    finishedBuilds,
-                    jobId: workflowGraph.nodes
-                        .find(node => node.name === trimJobName(nextJobName)).id
-                }));
-            }));
-        });
+                nextBuild = finishedInternalBuilds.filter(b => b.jobId === jobId)[0];
```
Comment (MysticDoll) [inline plugins/builds/index.js]:
```suggestion
                nextBuild = finishedInternalBuilds.find(b => b.jobId === jobId);
```

how about it?

File: plugins/builds/index.js [hunk 20]
```diff
@@ -293,83 +409,319 @@ exports.register = (server, options, next) => {
...
+
+                const newParentBuilds = deepmerge.all(
+                    [joinParentBuilds, currentJobParentBuilds,
+                        nextBuild.parentBuilds, currentBuildInfo]);
+
+                console.log('----newBuild updated', newParentBuilds);
+
+                nextBuild.parentBuilds = newParentBuilds;
+                // nextBuild.parentBuilds = deepmerge(parentBuilds, nextBuild.parentBuilds || {});
+                nextBuild.parentBuildId = [build.id].concat(nextBuild.parentBuildId || []);
```
Comment (MysticDoll) [inline plugins/builds/index.js]: `parentBuildIds` ?

File: plugins/builds/index.js [hunk 21]
```diff
@@ -293,83 +409,319 @@ exports.register = (server, options, next) => {
...
+                if (b.status === 'FAILURE' || b.status === 'ABORTED') hasFailure = true;
+                if (b.status !== 'SUCCESS' && b.status !== 'FAILURE') done = false; // some builds are still going on
+            });
+
+            /*  IF NOT DONE -> DO NOTHING
+                IF DONE ->
+                    CHECK IF HAS FAILURE -> DELETE NEW BUILD
+                    OTHERWISE -> START NEW BUILD
+               IF ALL SUCCEEDED -> START NEW BUILD
+           */
```
Comment (MysticDoll) [inline plugins/builds/index.js]: above 2 lines need 1 whitespace each.

File: plugins/builds/index.js [hunk 22]
```diff
@@ -293,83 +409,319 @@ exports.register = (server, options, next) => {
...
      * @param {Build}       config.build        Current build
      * @param {String}      config.username     Username
-     * @param {String}      config.scmContext   scm context
+     * @param {String}      config.scmContext   Scm context
+     * @param {Boolean}     config.externalJoin Flag to allow external join
      * @return {Promise}                        Resolves to the newly created build or null
      */
-    server.expose('triggerNextJobs', (config) => {
-        const { pipeline, job, build, username, scmContext } = config;
+    server.expose('triggerNextJobs', async (config) => {
```
Comment (MysticDoll) [inline plugins/builds/index.js]: This function is too large.
Could you separate this function into some functions?

File: plugins/builds/index.js (hunk 18 above)
Comment (tkyi) [reply plugins/builds/index.js]: looks like it doesn't work if `start` is `undefined`
left it as is for now

Comment (MysticDoll) [reply plugins/builds/index.js]: I see

File: plugins/builds/index.js (hunk 22 above)
Comment (DekusDenial) [reply plugins/builds/index.js]: agree. it's bit hard to follow and this requires some refactoring too
